src/transformStateWithClones.js

```diff
@@ -7,7 +7,25 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  actions.forEach((action) => {
+    let updatedState = { ...currentState };
+
+    if (action.type === 'addProperties') {
+      updatedState = { ...updatedState, ...action.extraData };
+    } else if (action.type === 'removeProperties') {
+      action.keysToRemove.forEach((key) => delete updatedState[key]);
```

Comment:
Ensure that `action.keysToRemove` is always an array to avoid runtime errors. Consider adding a check or validation for this.